warehouse/query-core/src/main/java/datawave/query/jexl/functions/EventFieldAggregator.java

```diff
@@ -0,0 +1,144 @@
+import org.apache.accumulo.core.data.ByteSequence;
+import org.apache.accumulo.core.data.Key;
+import org.apache.accumulo.core.data.Value;
+import org.apache.accumulo.core.iterators.SortedKeyValueIterator;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+public class EventFieldAggregator extends IdentityAggregator {
+    // speedy cache loading for types, duplicated from AttributeFactory with caching of types rather than classes
+    protected static LoadingCache<String,Type<?>> typeCache = CacheBuilder.newBuilder().maximumSize(128).expireAfterAccess(1, TimeUnit.HOURS)
+                    .build(new CacheLoader<String,Type<?>>() {
+                        @Override
+                        public Type<?> load(String clazz) throws Exception {
+                            Class<?> c = Class.forName(clazz);
+                            return (Type<?>) c.newInstance();
+                        }
+                    });
+    
+    private TypeMetadata typeMetadata;
+    private String defaultTypeClass;
+    
+    public EventFieldAggregator(String field, EventDataQueryFilter filter, int maxNextCount, TypeMetadata typeMetadata, String defaultTypeClass) {
+        super(Collections.singleton(field), filter, maxNextCount);
+        
+        this.typeMetadata = typeMetadata;
+        this.defaultTypeClass = defaultTypeClass;
+    }
+    
+    @Override
+    protected List<Tuple2<String,String>> parserFieldNameValue(Key topKey) {
+        String cq = topKey.getColumnQualifier().toString();
+        int nullIndex1 = cq.indexOf('\u0000');
+        String field = cq.substring(0, nullIndex1);
+        String value = cq.substring(nullIndex1 + 1);
+        
+        int dataTypeEnd = -1;
+        ByteSequence cf = topKey.getColumnFamilyData();
+        for (int i = 0; i < cf.getBackingArray().length; i++) {
+            if (cf.byteAt(i) == '\u0000') {
+                dataTypeEnd = i;
+            }
+        }
+        
+        if (dataTypeEnd <= 0) {
+            throw new RuntimeException("malformed key, cannot parse data type from event");
+        }
+        
+        String dataType = cf.subSequence(0, dataTypeEnd).toString();
+        
+        Set<String> normalizedValues = getNormalizedValues(dataType, field, value);
+        List<Tuple2<String,String>> fieldValuePairs = new ArrayList<>(normalizedValues.size());
+        for (String normalizedValue : normalizedValues) {
+            fieldValuePairs.add(new Tuple2<>(field, normalizedValue));
+        }
+        
+        return fieldValuePairs;
+    }
+    
+    @Override
+    protected ByteSequence parseFieldNameValue(ByteSequence cf, ByteSequence cq) {
+        ArrayList<Integer> nulls = TLD.instancesOf(0, cq, 1);
```

Comment:
Seems to me this whole method could be replaced with simply returning cq.  If you need a copy of the bytes then return new ArrayByteSequence(Arrays.copyOf(cq.getBackingArray(), cq.length()), 0, cq.length()).

Reply:
you are right, there is no reason to do this. I don't think there is any reason not to just return the cq.